community/kernel/src/main/java/org/neo4j/kernel/impl/newapi/PropertyCursor.java

```diff
@@ -177,6 +247,19 @@ private PropertyType type()
 
     @Override
     public Value propertyValue()
+    {
+        if ( txStateValue != null )
+        {
+            return txStateValue.value();
+        }
+
+        Value value = readValue();
+
+        assertOpen.assertOpen();
```

Comment:
Should we not `assertOpen` before we `readValue`?

Reply:
I though so too, but that leads to failures in TerminationOfSlavesDuringPullUpdatesIT, we did the same before here https://github.com/neo4j/neo4j/blob/3.4/community/kernel/src/main/java/org/neo4j/kernel/impl/api/store/StorePropertyCursor.java#L106. Somehow we read inconsistent data but are "saved" by checking the transaction before actually returning data. I never really understood, I think @burqen and @tinwelint knows more.

Reply:
Yes. So we are traversing a dynamic property chain, reading a long key or value that spans multiple records. If the property records are being deleted and reused while we are in the middle of reading such chain we can end up reading crap. To guard for this the slave terminate all transactions that fall outside of "safezone" (are too old) when pulling updates from master. But if we don't check this termination flag before returning that does not help us and checking it before is not enough ofc because the termination and record reuse can happen after we start reading.